### ISSUE TITLE
Mayriad's LRR Bugs and Suggestions

### ISSUE THREAD

**Issue by Mayriad** [Difegue/LANraragi#274]
Firstly, thanks a lot for making this fantastic library manager. I think I have done a good amount of testing and found enough issues, so I will open a big thread here. I cannot come up with a clever title so I just used my standard format.

I will edit this if I find new problems or suggestions to add. A lot of suggestions can be made, but I will focus on bugs for now.

## 1. Index

Bugs:

- [ ] 4.1. Plugins.pm missing the level after "::Plugin" in the namespace
- [ ] 4.2. Plugin update not effective until LANraragi is restarted
- [ ] 4.3. Plugin upload not working after failed upload
- [ ] 4.4. Admin settings not saving
- [ ] 4.5. Cleaning and resetting database currently broken
- [ ] 4.6. Shinobu permanently omitting files after it stops half way
- [ ] 4.7. Shinobu not responding to new archives if they are added in folders
- [ ] 4.8. Shinobu adding new archives twice if they are not added in folders
- [ ] 4.9. Batch tagging still throwing code 1006
- [ ] 4.10. Checkboxes too tiny on the batch tagging page
- [ ] 4.11. Undefined value error from the E-Hentai metadata plugin

Suggestions:

- [ ] 5.1. Namespace shorthand and naming
- [ ] 5.2. More options for the E-Hentai plugin
- [ ] 5.3. Better gallery lookup routine for the E-Hentai plugin
- [ ] 5.4. Different treatment for the dates from the DateAdded plugin
- [ ] 5.5. Delay between requests for auto-plugin
- [ ] 5.6. Empty tag blacklist
- [ ] 5.7. LANraragi thread in EH forums

## 2. Setup

LANraragi for Windows 0.6.9 (MSI installer) on Microsoft Windows 10 version 10.0.18363.720. The LANraragi content folder is on the Windows PC, but the archives are stored on a NAS and accessed via a WSL mount point in the content folder. The PC and the NAS are connected via WIFI 5. My NAS cannot run docker, and Homebrew is either impossible or too difficult for me, so I installed LANraragi on Windows. Maybe some of the bugs below are caused by my setup, but I cannot know.

## 3. Myself

I just started using LANraragi to organise galleries downloaded from EH. I never studied ComSci and do not know Linux and Perl at all before using LANraragi, but I have programming experience and can hence read the code to guess how things work. I can either fix or avoid the problems below, and I have also written my own metadata plugin to use with the next update of [my script](https://openuserjs.org/scripts/Mayriad/Mayriads_EH_Master_Script), so I am not really complaining.

### 4.1. Plugins.pm missing the level after "::Plugin" in the namespace

The subroutine `process_upload()` in `/lib/LANraragi/Controller/Plugins.pm` incorrectly assumes all uploaded plugins have the namespace `LANraragi::Plugin::PluginName`, and does not take into account three possible types of plugins: login, metadata and scripts. This breaks the "upload script" functionality and LANraragi will always say:
```
[Plugin Upload] [error] Could not instantiate plugin at namespace LANraragi::Plugin::MyNewPlugin!
[Plugin Upload] [error] Can't locate object method "plugin_info" via package "LANraragi::Plugin::MyNewPlugin" (perhaps you forgot to load "LANraragi::Plugin::MyNewPlugin"?) at /home/koyomi/lanraragi/script/../lib/LANraragi/Controller/Plugins.pm line 168.
```

Line 151 and 162 can be directly edited to fix this problem and allow one type of plugin to be uploaded, which is what I did. For example, to add the example plugin `MyNewPlugin.pm` from the documentation, which is a metadata plugin, these lines need to be changed from:
```
my $dir = getcwd() . ("/lib/LANraragi/Plugin/");
...
my $pluginclass = "LANraragi::Plugin::" . substr( $filename, 0, -3 );
```
To:
```
my $dir = getcwd() . ("/lib/LANraragi/Plugin/Metadata/");
...
my $pluginclass = "LANraragi::Plugin::Metadata::" . substr( $filename, 0, -3 );
```

### 4.2. Plugin update not effective until LANraragi is restarted

I am not sure about this one. When I uploaded a new version of an existing plugin to LANraragi using the "upload plugin" button, I think I did not see the changes until I restarted LANraragi. When a plugin is uploaded, the message says "refresh the page to see it," so I expect the update to be effective without having to restart the program.

### 4.3. Plugin upload not working after failed upload

After a plugin is uploaded to LANraragi using the "upload plugin" button but fails due to some problem like syntax error, uploading the fixed, working version of the same plugin can lead to this error popup:

```
Error uploading plugin
Could not load namespace LANraragi::Plugin::Metadata::*! Your Plugin might not be compiling properly.
Here's an error log: Attempt to reload /home/koyomi/lanraragi/lib/LANraragi/Plugin/Metadata/*.pm aborted. Compilation failed in require at /home/koyomi/lanraragi/script/../lib/LANraragi/Controller/Plugins.pm line 167.
```

This error will not go away until LANraragi is restarted i.e., after a reboot the same plugin will be successfully uploaded, but before that uploading this plugin will always give this error.

### 4.4. Admin settings not saving

Sometimes clicking the "save settings" button will give a popup that says "error while saving: internal server error". This problem can be reproduced by spamming the save button. In debug mode these seem to be the relevant lines:

```
[Mojolicious] [7ab7e0a9] Error while reading from Redis server:  at /home/koyomi/lanraragi/script/../lib/LANraragi/Controller/Config.pm line 130.
[Mojolicious] Template "exception.development.html.tt2" not found
[Mojolicious] Template "exception.html.tt2" not found
[Mojolicious] [7ab7e0a9] Rendering cached template "mojo/debug.html.ep"
[Mojolicious] [7ab7e0a9] 500 Internal Server Error (0.027213s, 36.748/s)
[Mojolicious] [cab544ca] POST "/config"
[Mojolicious] [cab544ca] Routing to controller "LANraragi::Controller::Login" and action "logged_in"
[Mojolicious] [cab544ca] Routing to controller "LANraragi::Controller::Config" and action "save_config"
```

### 4.5. Cleaning and resetting database currently broken

I think you know this already. Pressing these buttons will give error messages and puking dinosaurs in the GUI but nothing in the console. Not sure if it is relevant, but in debug mode this message will be given:

```
Can't locate object method "FIRSTKEY" via package "Parallel::Loops::TiedHash" at /home/koyomi/lanraragi/script/../lib/LANraragi/Utils/Generic.pm line 71.
```

### 4.6. Shinobu permanently omitting files after it stops half way

When the Shinobu background worker imports a lot of galleries at once, it can stop running half way through without throwing anything. Then the background worker will need to be restarted to continue the addition. This can happen fairly frequently and Shinobu can stop twice while importing 1,000 archives. When Shinobu stops, its status is still "OK!" in the admin settings and no error is given even in debug mode.

Shinobu will eventually say "all done!" after necessary restarts, but it is possible that a few archives are not added at all and the total counts would differ between LANraragi and my file system. These archives are permanently omitted because moving them around and restarting stuff cannot trigger Shinobu to add them. I think this problem can always happen and is not affected by auto-tag and auto-plugin.

This is a serious bug, but it happens randomly and so far I can only report three findings; after Shinobu stops:

- The GUI can lag a bit.
- Untagged archives will not be automatically ticked on the batch tagging page.
- Batch tagging will start slowly, so after the "Started Batch Tagging operation..." message. it can take quite a while before it tags the first archive.

### 4.7. Shinobu not responding to new archives if they are added in folders

As mentioned in section 2, LANraragi watches and accesses my archives via a mount point in the content folder. However, no matter where I add the archives, as long as I move them into the content folder in folders, Shinobu will not respond to these new archives in folders even though its status is still "OK!" in the admin settings. There is no log to provide.

So, Shinobu will not react if I add archives in folders; Shinobu does react if I add the same archives to the same place without placing them inside folders, but another problem will happen. See the next bug below.

### 4.8. Shinobu adding new archives twice if they are not added in folders

Continuing from the previous section, no matter where I add the archives, as long as I move them into the content folder directly without using folders, according to the debug mode messages it seems Shinobu actually adds these archives twice and gives a duplicate warning that basically says "this path is a duplicate of exactly the same path":

```
[Shinobu] [debug] Received inotify event create on /mnt/d/System/Lanraragi/Galleries/*.zip
[Shinobu] [debug] Adding /mnt/d/System/Lanraragi/Galleries/*.zip to Shinobu filemap.
[Shinobu] [debug] Computed ID is 3d11d35145bcf98503ab66bea76fbe06a3ed9819.
[Shinobu] [info] Adding new file /mnt/d/System/Lanraragi/Galleries/*.zip with ID 3d11d35145bcf98503ab66bea76fbe06a3ed9819
[Archive] [debug] Pushing to redis on ID 3d11d35145bcf98503ab66bea76fbe06a3ed9819:
[Archive] [debug] File Name: *
[Archive] [debug] Filesystem Path: /mnt/d/System/Lanraragi/Galleries/*.zip
[Shinobu] [debug] Received inotify event modify on /mnt/d/System/Lanraragi/Galleries/*.zip
[Shinobu] [debug] Adding /mnt/d/System/Lanraragi/Galleries/*.zip to Shinobu filemap.
[Shinobu] [debug] Computed ID is 3d11d35145bcf98503ab66bea76fbe06a3ed9819.
[Shinobu] [warn] /mnt/d/System/Lanraragi/Galleries/*.zip is a duplicate of the existing file /mnt/d/System/Lanraragi/Galleries/*.zip. You should delete it.
```

### 4.9. Batch tagging still throwing code 1006

This was reported and fixed in [issue 153](https://github.com/Difegue/LANraragi/issues/153) but I still have the same problem and I am not importing 47,000 archives like the other user. I think in my case it is most likely to happen when the connection between my PC and NAS is busy. If I transfer files from my PC to my NAS while batch tagging is tagging archives on the NAS, most of the time code 1006 will be quickly thrown; however, this problem can still occur when I do not transfer files. It can happen randomly and is not affected by:

- What plugin is used to tag the archives, because the E-Hentai plugin and my own plugin can both throw code 1006.
- Whether the option "fetch using thumbnail first" in the E-Hentai plugin was used i.e., the search method does not matter.
- Whether LANraragi makes any modification to metadata. This problem still occurred when I ran batch tagging over and over on the same archives without adding any new metadata.

When it happens, a blue bubble that says "batch tagging complete" will pop up even though it ended abnormally, and the log in the batch tagging GUI does not have any useful information:

```
...
Processed *(Added tags:  *)
Sleeping for * seconds.
************
(code 1006)
```

The log console seems to have some relevant messages:

```
[E-Hentai] [info] Sending the following tags to LRR: *
[Batch Tagging] [debug] No user overrides given.
[Batch Tagging] [debug] Processing *
[Plugin System] [info] Thumbnail hash invalid, regenerating.
[Mojolicious] Worker 43 has no heartbeat (30 seconds), restarting
[Mojolicious] Stopping worker 43 gracefully (120 seconds)
[Mojolicious] Worker 1184 started
[Mojolicious] Stopping worker 43 immediately
[Mojolicious] Worker 43 stopped
```

The console stops here and does not give anything else.

### 4.10. Checkboxes too tiny on the batch tagging page

When "override plugin global arguments" is enabled, the CSS for the `table.tag-options` element containing the `input.stdinput` checkboxes need to have their width specified. Right now their width can be too low and consequently the proportionally sized checkboxes can be nearly invisible. Most of the plugins are affected by this visual problem on the batch tagging page.

It can be fixed by adding a width property to `table.tag-options` or somewhere else. Currently it only has an inherited width property that is not applicable.

### 4.11. Undefined value error from the E-Hentai metadata plugin

I noticed one gallery had this problem in the batch tagging log and it could not be processed:

```
...
Sleeping for * seconds.
Plugin error while processing ID *(Can't use an undefined value as a HASH reference at /home/koyomi/lanraragi/script/../lib/LANraragi/Plugin/Metadata/EHentai.pm line 229.
)
Sleeping for * seconds.
...
```

This happened while I was using the E-Hentai plugin with the option "fetch using thumbnail first" enabled. I cannot provide more information because I think it has only happened once and I have no idea which archive was affected among the hundreds tagged.

### 5.1. Namespace shorthand and naming

Currently LANraragi does not support namespace shorthand in search terms, which means terms like `l:english` will not work and you have to type `language:english`. Allowing shorthand like `l:` can make it much easier to type these search terms. I understand the difficulty here since LANraragi allows custom namespaces to be created freely and also handles metadata from different websites that may use different namespaces, but this problem can be solved by abbreviating popular namespaces only e.g., `p:` will only mean `parody:` and not a custom namespace like `posted:`.

The EH metadata plugin I wrote for my script retrieves additional timestamp and uploader metadata and I have been assuming that eventually LANraragi can use the abbreviations below:
```
a - artist
c - character
f - female
g - group
m - male
l - language
r - reclass
p - parody

u - uploader
t - timestamp ("posted" in JSON returned by EH API)
i - imported (replacing "date_added" from the DateAdded metadata plugin)
d - division (replacing the default "category" namespace)
```
This way probably all namespaces can be abbreviated, but of course this is just my wishful thinking.

### 5.2. More options for the E-Hentai plugin

Some more options can be easily added to this plugin. My plugin has the first two options below; the third one is not necessary because my plugin does not search for galleries.

- Option to save the Japanese gallery title, which is the "title_jpn" in the JSON from EH API. Right now the plugin only has an option to save the English or romanised title, but I am sure there are users like me who can read or prefers to preserve the original Japanese title.

- Option to save additional uploader and timestamp (time posted) metadata from the JSON from EH API. `uploader:` is effectively a searchable namespace on EH and also useful in LANraragi. `posted:` is named `timestamp:` in my plugin to let it have its own namespace shorthand `t:`; this namespace will be useful when LANraragi supports advanced search options in the future.

- Option to search and match expunged galleries as well. Adding `&fs_exp=1` to line 121 in `EHentai.pm` should enable it. It is definitely not a good idea to enable this everyday, but it could be enabled on demand to help people who are archiving systematically expunged galleries like `group:lilith`.

### 5.3. Better gallery lookup routine for the E-Hentai plugin

I think the subroutine `lookup_gallery()` in the E-Hentai plugin can use more and better steps to find the source. Right now, depending on `$usethumbs`, it is either a title search using the filename or a reverse image search using the cover. It is too prone to no-match and mismatch.

I have not really tested this issue and any alternatives, because I have been using and testing my own metadata plugin, but I think at least the plugin can use two steps, where it will try the image search for archives left unmatched after title search.

### 5.4. Different treatment for the dates from the DateAdded plugin

The dates added by the DateAdded plugin are tags, so all galleries will become "tagged" when this plugin is enabled, even if some of them are not linked to source and have no real metadata. This plugin will therefore hide truly untagged archives that require manual tagging.

I am not sure if you can do something about this now, but I think in the long term dates will need to be stored and handled differently from tags (along with the `timestamp:` I mentioned above), so that they can be actually useful in searches and sorting of archives.

### 5.5. Delay between requests for auto-plugin

I think the auto-plugin does not have a delay between metadata requests? Adding 1,000+ galleries would probably lead to a temporary ban. It should be safer and nicer to websites to build a delay between requests or a longer break after tagging a number of archives.

### 5.6. Empty tag blacklist

For some reason, the tag blacklist in the admin settings cannot be empty, which means LANraragi requires the user to omit some tag(s). I do not wish to omit anything, so I just put down "none", which is not a real tag.

### 5.7. LANraragi thread in EH forums

You should consider making a LANraragi thread in [the E-Hentai forums](https://forums.e-hentai.org/index.php?showforum=7) so that more people will get to know and try it, unless you are too busy to maintain a thread there. The resident moderator penguin will probably pin your thread right away considering how good LANraragi is. The only problem is that we have too many pinned threads in the site discussion and helpdesk subforum.

**Comment by Difegue** [comment]
Alright, first things first don't take it personally but I'm going to close this issue:  

Issues are meant to be laser focused on one problem/enhancement request, and stashing multiple problems in one makes it **really hard** for both me and other users to track.  
If multiple people have the same problem/enhancement request, they can look at the specific issue about it and comment/+1 it, making it easier for me to prioritize what to work on as well.

I've let it slide [on other](https://github.com/Difegue/LANraragi/issues/114) [occasions](https://github.com/Difegue/LANraragi/issues/116), but your one issue does take it a magnitude higher. 😅 I'll be closing the other ones too.  

Since most of your points are very valid I'll make a broad reply below about all of them -- Please read it and reopen single issues tackling one problem at a time.

![73d9b5e9ec7be383bfc02154fba942c484620c8a4af4b580fc58e19c96e245de](https://user-images.githubusercontent.com/8237712/79902693-33b50300-8412-11ea-835b-c5d6e4223468.jpg)

**Comment by Difegue** [comment]
### 4.1
Indeed, I forgot to fix the plugin uploading code when I introduced the three types in 0.6.9.  
This code doesn't get used too often, as I don't have many plugin developers 😓.  
Since it seems you already made some headway towards fixing it, it'd be great if you could finish this up and make a PR for it. I'll work on it otherwise, but I might take a while to get to it. 😊  

### 4.2
"I am not sure about this one." -> Make a proper repro and open a new issue if the problem is real.

### 4.3
Might be related to the fix for 4.1, worth opening a separate issue for.

### 4.4
Sounds like a redis overloading problem by trying too many saves at once. Not critical, won't fix.

### 4.5
Fixed in 0.6.9-EX.

### 4.6
I've had a few reports of a similar bug, but so far I have no idea what causes it. I thought the recent multiprocessing changes would fix it, but if you're running 0.6.9 I guess it doesn't do enough.  
I'm guessing it's a load problem where the server gets overwhelmed and fucks up the Shinobu process, but since it spits no errors I haven't been able to give it a proper diagnosis.
I'd like a separate issue open for this as well so it can be properly tracked.

### 4.8
"Fixed" in 0.6.9-EX: This is another problem with WSL's inotify implementation, where it's counting two events. Shinobu never adds archives twice (it can't as the ID is the same), but the message was misleading and has been changed.

### 4.9
From a user on Discord: 
> if you didn't already fix error 1006 yourself, you can do it by setting the environment variable MOJO_KEEP_ALIVE_TIMEOUT=25
I should probably add this to the default for both Docker and Windows.

### 4.10
Please make a PR with the fix if you can. 😀

### 4.11
Would need a proper reproduction case before I can attempt a fix.

### 5.1
I've had a request of the like already which I think is buried in the "Usability Improvements" issue (another example as to why issues with multiple items are very hard to track and act on). 
Please make a separate "enhancement" issue. 

### 5.2 
Please make a separate "enhancement" issue as well. The E-H plugin is already rather complex as-is so I probably won't tackle this in the near future.

### 5.3
Ditto.

### 5.4 
IIRC the "look for untagged archives" functions already ignore the dateadded namespace for this exact reason. If you have a clean reproduction case please file a separate bug issue.
I don't plan on modifying LRR's data model at the time, which means no extra storage fields for archives.

### 5.5
There's a small hardcoded delay I believe, might be worth digging further.

### 5.6
Real bug, please make a separate issue so I can track and fix it

### 5.7
I mostly only use the Discord server nowadays to discuss with the userbase -- It's easy to access, centralizes information and I personally consider that the software is well-known enough that I don't need to bother with marketing that much anymore 😉:  
![image](https://user-images.githubusercontent.com/8237712/79901766-9f966c00-8410-11ea-8f5f-38438f265849.png)  
I don't mind if other people try to evangelize/spread the word about it, though -- Might make for more donations!
![fe2785a206daeec125fc84e482fadc13fd344b75424943abc3ec8e6a4774d9d0](https://user-images.githubusercontent.com/8237712/79902313-91951b00-8411-11ea-88f4-81b3fbc5f2a6.gif)

**Comment by Mayriad** [comment]
Thank you for your prompt reply. I do not mind you closing the issues at all, since I have got your answers. I knew such a huge thread would be difficult to track, but I thought it would not be nice to spam issues, and I could connect different sections in a huge thread. Anyway, at least this thread helped to screen out some issues first.

So, these issues have been fixed:
- 4.5. Cleaning and resetting database currently broken
- 4.8. Shinobu adding new archives twice if they are not added in folders

These issues can be ignored:
- 4.4. Admin settings not saving
- 5.7. LANraragi thread in EH forums

I think I can open PRs to address these when I get time:
- 4.1. Plugins.pm missing the level after "::Plugin" in the namespace
- 4.10. Checkboxes too tiny on the batch tagging page
- 5.2. More options for the E-Hentai plugin

Before I talk about these again, I will need to test them first when I get even more time:
- 4.2. Plugin update not effective until LANraragi is restarted
- 4.6. Shinobu permanently omitting files after it stops half way
- 4.9. Batch tagging still throwing code 1006
- 4.11. Undefined value error from the E-Hentai metadata plugin

Lastly, I will open separate issues for the remaining ones:
- 4.3. Plugin upload not working after failed upload
- 4.7. Shinobu not responding to new archives if they are added in folders (you did not answer this one, so I assume I should open an issue)
- 5.1. Namespace shorthand and naming
- 5.3. Better gallery lookup routine for the E-Hentai plugin
- 5.4. Different treatment for the dates from the DateAdded plugin (no, it does not ignore "data_added")
- 5.5. Delay between requests for auto-plugin
- 5.6. Empty tag blacklist

I will spam 7 issues in total.
